segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -704,11 +710,152 @@ public void readTable(final WireCommands.ReadTable readTable) {
+                                                                                                                           keyArray.getLength()), k.getVersion());
+                                                                        })
+                                                                        .collect(toList());
+                      connection.send(new WireCommands.TableKeysRead(readTableKeys.getRequestId(), segment, wireCommandKeys, continuationToken.get()));
+                  }).exceptionally(e -> handleException(readTableKeys.getRequestId(), segment, operation, e));
+    }
+
+    @Override
+    public void readTableEntries(WireCommands.ReadTableEntries readTableEntries) {
+        final String segment = readTableEntries.getSegment();
+        final String operation = "readTableEntries";
+
+        if (!verifyToken(segment, readTableEntries.getRequestId(), readTableEntries.getDelegationToken(), operation)) {
+            return;
+        }
+
+        log.info(readTableEntries.getRequestId(), "Fetching keys from {}.", readTableEntries);
+
+        int suggestedEntryCount = readTableEntries.getSuggestedEntryCount();
+        ByteBuf token = readTableEntries.getContinuationToken();
+
+        byte[] state = null;
+        if (!token.equals(EMPTY_BUFFER)) {
+            state = token.array();
+        }
+
+        final AtomicInteger msgSize = new AtomicInteger(0);
+        final AtomicReference<ByteBuf> continuationToken = new AtomicReference<>(EMPTY_BUFFER);
+        final List<TableEntry> entries = synchronizedList(new ArrayList<>());
+        tableStore.entryIterator(segment, state, TIMEOUT)
+                  .thenCompose(itr -> itr.collectUntil(
+                          e -> {
+                              final Collection<TableEntry> tableEntries = e.getEntries();
+                              final ArrayView lastState = e.getState();
+
+                              // Store all TableEntrys.
+                              entries.addAll(tableEntries);
+                              // Update the continuation token.
+                              continuationToken.set(wrappedBuffer(lastState.array(), lastState.arrayOffset(), lastState.getLength()));
+                              // Update message size.
+                              msgSize.addAndGet(getTableEntryBytes(segment, tableEntries, lastState.getLength()));
+
+                          },
+                          e -> entries.size() < suggestedEntryCount && msgSize.get() < MAX_READ_SIZE))
+                  .thenAccept(v -> {
+                      log.debug(readTableEntries.getRequestId(), "{} entries obtained for ReadTableEntries request.", entries.size());
+                      List<Map.Entry<WireCommands.TableKey, WireCommands.TableValue>> wireCommandEntries =
+                              entries.stream()
```

Comment:
Unfortunately, Java decided to be totally dumb. This doesn't work. The stream method on a synchronized list is not synchronized. (I know)

Reply:
Thanks for pointing it. Updated the code.